File: source/streams/src/parsers/basic.parser.ts
import { ChatMessage } from "@shared/models";
import IStreamParser from "./interface";

export default class BasicParser implements IStreamParser<ChatMessage> {
  parse(raw: Record<string, string>) {
    if (!raw.message) return null;

    const { message, ...remains } = raw;

    try {
      const payload = JSON.parse(message);

      return {
        eventName: "receive_message",
        payload,
        ...remains,
      };
    } catch (err) {
      console.error("[RedisStreamParser] Failed to parse record.message", err);
      return null;
    }
  }
}

File: project/consumers/source/streams/src/main/scala/parser/BasicParser.scala
package consumers.streams.source
package parser

import scala.util.{Try, Success, Failure}

import play.api.libs.json._

import consumers.shared.model.ChatMessage

class BasicParser extends IStreamParser[ChatMessage] {

  implicit val chatMessageReads: Reads[ChatMessage] = Json.reads[ChatMessage]

  def parse(raw: Map[String, String]): Option[ParsedEvent[ChatMessage]] = {
    raw.get("message") match {
      case None => None
      case Some(message) =>
        Try(Json.parse(message).as[ChatMessage]) match {
          case Success(payload) =>
            Some(
              ParsedEvent(
                eventName = "receive_message",
                payload = Right(payload),
                `type` = raw.get("type"),
                nsp = raw.get("nsp"),
                uid = raw.get("uid")
              )
            )
          case Failure(err) =>
            println(s"[RedisStreamParser] Failed to parse record.message: $err")
            None
        }
    }
  }
}
